### ISSUE TITLE
Workflow when using versioneer with GitHub repositories

### ISSUE THREAD

**Issue by danieljacobs1** [warner/python-versioneer#159]
I've been investigating use of `versioneer` for our projects at work, but I think I have run into an issue that maybe someone can provide some insight on.

We use GitHub for our projects, and we use a branch named `dev` as an integration branch for our development. Developers branch from `dev` for bug and feature work, and merge back into it once their work is completed.

Here's a rough sketch of that (`o` represents a commit):

```
tags                    v1.2.0  v1.2.1
                            |    |
master    ------------------o----o-----
                           /     /
                          /     /
dev       o--------------o-----o-------
           \            / \   /
feature 1   o----o-o-o-o   \ /
bug fix 1                   o 
```

When we want to make a release, we create a pull request from `dev` into `master`. Usually we use the standard "Merge pull request" (as opposed to the "Squash and merge" or "Rebase and merge" options on GitHub).

The "Merge pull request" uses git's `--no-ff` option, meaning that a commit merge is created on `master` for each release.

We tag the merge commit on `master` to indicate that that is the release commit.

The problem I have with `versioneer` is that when I have checked out `dev` (say in the scenario sketched above), it doesn't find any tags at all and it reports version numbers such as `0+untagged.40.g79cbd06` (based on the `setup.cfg` of course).

From experimentation, I can see that if I merge directly into `master` without `--no-ff` option it seems to work as intended. My guess is that it is because there is a common commit in the history of `master` and `dev` branches. However, GitHub actually prevents such common commits.

So my questions are:

- Is this the expected behaviour for `versioneer`?
- If it is expected behaviour, how can we use `versioneer` with GitHub? Am I doing something wrong?
- Is there a recommended change management plan/branching strategy that works with `versioneer` and GitHub?

Thanks in advance.

**Comment by jayvdb** [comment]
> The problem I have with versioneer is that when I have checked out dev (say in the scenario sketched above), it doesn't find any tags at all and it reports version numbers such as 0+untagged.40.g79cbd06 (based on the setup.cfg of course).

My guess is you are pull `dev` from your fork, and not from the origin?
In which case, the tags probably dont exist on your fork, because they are not cloned/forked by default.

**Comment by danieljacobs1** [comment]
> My guess is you are pull dev from your fork, and not from the origin?

I don't think that's it as I have pulled from the origin.

**Comment by jcfr** [comment]
After each release, you could merge `master` back into `dev`

**Comment by danieljacobs1** [comment]
> After each release, you could merge `master` back into `dev`.

We're using git at the command line to do some of the merges and that works very well. Thanks all the same.

**Comment by Juanlu001** [comment]
I came here when looking how to make versioneer work with a "trunk-based workflow":

https://www.toptal.com/software/trunk-based-development-git-flow

**Comment by danieljacobs1** [comment]
> Isn't this a simplified (no release branch) version of nvie/gitflow#49?

Possibly, I'm not too familiar with git flow.

>  It seems to me quite proper that tools wouldn't look at 'cousins' for tags to identify the current commit.

That's fair enough and quite reasonable. I just wanted to make sure I hadn't missed something obvious.

Just for posterity, in the end we stopped using GitHub for merging, and instead used the `git` command line which gave the desired effect.

**Comment by altendky** [comment]
I'm not sure what difference GitHub vs. git would give you (other than both having multiple options on how to merge and perhaps you picking a different option).  Maybe you are now fast-forward merging and your master branch head just now points at the same commit as dev?  If you really want a separate release branch then I expect you'll end up wanting to do a (`--no-ff`) merge from master back into dev.

**Comment by danieljacobs1** [comment]
> Maybe you are now fast-forward merging and your master branch head just now points at the same commit as dev?

Yes, that's what we did, but things have moved on since then for us.
